plugins/gradle/java/testSources/compiler/GradleImprovedHotswapDetectionTest.java

```diff
@@ -0,0 +1,373 @@
+      implJar
+    );
+
+    // note that the "implJar" is not marked as a generated file
+    // this is jar itself is marked as dirty root
+    assertThat(generatedFiles).as("Generated files").containsOnly(
+      Map.entry(implMainRoot, Set.of("my/pack/Impl.class"))
+    );
+  }
+
+  @Test
+  public void testBuildTestProjectAfterMain() {
+    compileModules("project.main");
+
+    clearOutputs();
+    compileModules("project.test");
+
+    assertThat(dirtyOutputRoots).containsExactlyInAnyOrder(testRoot);
+    assertThat(generatedFiles).as("Generated files").containsOnly(
+      Map.entry(testRoot, Set.of("my/pack/AppTest.class"))
+    );
+
+    assertCopied(testRoot + "/my/pack/AppTest.class");
+    assertNotCopied(apiTestRoot + "/my/pack/ApiTest.class");
+    assertNotCopied(implTestRoot + "/my/pack/ImplTest.class");
+  }
+
+  @Test
+  public void testRebuildMainProjectAfterUndoingChange() {
+    compileModules("project.main");
+
+    setFileContent(appFile, "package my.pack;\n" +
+                            "public class App {\n" +
+                            "  public int method() { return 42; }\n" +
+                            "  public int methodX() { return 42; }\n" +
+                            "}", false);
+
+    clearOutputs();
+    compileModules("project.main");
+
+    // revert file to previous value
+    setFileContent(appFile, "package my.pack;\n" +
+                            "\n" +
+                            "public class App {\n" +
+                            "  public int method() { return 42; }\n" +
+                            "}", false);
+
+    clearOutputs();
+    compileModules("project.main");
+
+    assertThat(dirtyOutputRoots).containsExactly(mainRoot);
+    assertThat(generatedFiles).containsOnly(Map.entry(mainRoot, Set.of("my/pack/App.class")));
+  }
+
+  @Test
+  public void testBuildProjectWithResources() throws IOException {
+    compileModules("project.main");
+    createProjectSubFile("src/main/resources/runtime.properties",
+                         "resourceString=foobar");
+
+    clearOutputs();
+    compileModules("project.main");
+
+    assertThat(dirtyOutputRoots).containsExactly("build/resources/main");
+    assertThat(generatedFiles).containsOnly(Map.entry("build/resources/main", Set.of("runtime.properties")));
+
+  }
+
+  private void subscribeToProject() {
+    MessageBusConnection connection = myProject.getMessageBus().connect(getTestRootDisposable());
+    connection.subscribe(ProjectTaskListener.TOPIC, new ProjectTaskListener() {
+      @Override
+      public void started(@NotNull ProjectTaskContext context) {
+        context.enableCollectionOfGeneratedFiles();
+      }
+
+      @Override
+      public void finished(@NotNull ProjectTaskManager.Result result) {
+        result.getContext().getDirtyOutputPaths()
+          .ifPresent(paths -> dirtyOutputRoots.addAll(paths.map(PathUtil::toSystemIndependentName).map(path -> relativePath(path)).collect(Collectors.toList())));
```

Comment:
maybe a multiline internal lambda instead of it all getting thrown in `addAll` ?